content/automate/ManageIQ/Service/Retirement/StateMachines/Methods.class/__methods__/check_service_retired.rb

```diff
@@ -1,37 +1,71 @@
-#
-# Description: This method checks to see that all of the service resources are retired before retiring the service.
-#
-
-service = $evm.root['service']
-if service.nil?
-  $evm.log('error', "Service Object not found")
-  exit MIQ_ABORT
-end
+module ManageIQ
+  module Automate
+    module Service
+      module Retirement
+        module StateMachines
+          module Methods
+            class CheckServiceRetired
+              def initialize(handle = $evm)
+                @handle = handle
+              end
+
+              def main
+                service = service_obj
+
+                @handle.log('info', "Checking if all service resources have been retired.")
+                result = 'ok'
+                service.service_resources.each do |sr|
+                  next if sr.resource.nil?
+                  next if sr.resource_type == "Service" && @handle.vmdb(:service).find(sr[:resource_id]).has_parent
+                  next unless sr.resource.respond_to?(:retired?)
+                  result = check_service_resource_retirement(sr, service)
+                end
+
+                @handle.log('info', "Service: #{service.name} Resource retirement check returned <#{result}>")
+                check_retired_state(result, service)
+              end
+
+              private
 
-$evm.log('info', "Checking if all service resources have been retired.")
+              def service_obj
```

Comment:
@d-m-u 
```
 def service
    @service ||= @handle.root['service'].tap do |obj|
                           if obj.nil?
                               @handle.log(:error, 'Service object not provided')
                               raise 'Service object has not been provided'
                            end
                         end
                  end
```
If you did it this way it will cache the service object in an instance variable once after all the validation and for subsequent calls you can just use service every where.

You can remove the
 service = service_obj line
Not pass the service into check_retired_state